src/auto_scheduler/search_policy/sketch_policy_rules.h

```diff
@@ -73,84 +73,51 @@ class SketchGenerationRule {
                                                    const State& state, int stage_id) const = 0;
 };
 
+#define DEFINE_SKETCH_GENERATION_RULE(rule_name)                                                 \
```

Comment:
I'd like to raise a discussion about using macro here. Although we can of course largely reduce the code size with this macro, it makes code tracing more difficult. For example, most IDE navigation cannot locate `class RuleSkipStage` anymore. If we want to avoid redundant lines of 

```
std::vector<std::pair<State, int>> Apply(const SketchPolicyNode& policy, const State& state,	
                                           int stage_id) const final;
```

maybe we can make it not a virtual function so that we can simply override it in the cc file without declaration in .h file?

Reply:
I don't have clear idea on this part, but I think it's fine to try your suggestion.
Simply merged this PR first, so it will be more convenient for you to try different class structures in your PR.